include/wil/registry_helpers.h

```diff
@@ -0,0 +1,947 @@
+                {
+                    return true;
+                }
+                inline HRESULT prepare_buffer(::std::wstring& string) WI_NOEXCEPT
+                {
+                    // zero the input string if there is space already allocated
+                    for (auto& string_char : string)
+                    {
+                        string_char = L'\0';
+                    }
+                    return S_OK;
+                }
+                inline void* get_buffer(const ::std::wstring& string) WI_NOEXCEPT
+                {
+                    return const_cast<wchar_t*>(string.data());
+                }
+#endif // #if defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+                constexpr void* get_buffer(const int32_t& value) WI_NOEXCEPT
+                {
+                    return const_cast<int32_t*>(&value);
+                }
+
+                constexpr void* get_buffer(const uint32_t& value) WI_NOEXCEPT
+                {
+                    return const_cast<uint32_t*>(&value);
+                }
+
+                constexpr void* get_buffer(const long& value) WI_NOEXCEPT
+                {
+                    return const_cast<long*>(&value);
+                }
+
+                constexpr void* get_buffer(const unsigned long& value) WI_NOEXCEPT
+                {
+                    return const_cast<unsigned long*>(&value);
+                }
+
+                constexpr void* get_buffer(const int64_t& value) WI_NOEXCEPT
+                {
+                    return const_cast<int64_t*>(&value);
+                }
+
+                constexpr void* get_buffer(const uint64_t& value) WI_NOEXCEPT
+                {
+                    return const_cast<uint64_t*>(&value);
+                }
+
+                constexpr void* get_buffer(PCWSTR value) WI_NOEXCEPT
+                {
+                    return const_cast<wchar_t*>(value);
+                }
+
+#if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+                inline void* get_buffer(const ::std::vector<wchar_t>& value) WI_NOEXCEPT
+                {
+                    return const_cast<wchar_t*>(value.data());
+                }
+#endif // #if defined(_VECTOR_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+#if defined(__WIL_OLEAUTO_H_)
+                inline void* get_buffer(const BSTR& value) WI_NOEXCEPT
+                {
+                    return value;
+                }
+                inline void* get_buffer(const ::wil::unique_bstr& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // #if defined(__WIL_OLEAUTO_H_)
+
+#if defined(__WIL_OLEAUTO_H_STL)
+                inline void* get_buffer(const ::wil::shared_bstr& value) WI_NOEXCEPT
+                {
+                    return value.get();
+                }
+#endif // #if defined(__WIL_OLEAUTO_H_STL)
+
+#if defined(__WIL_OBJBASE_H_)

```

Comment:
Ok, this "buffer" thing looks an awful lot like the string_maker type. Consider using that one instead.

Reply:
Or rather, I _thought_ there was a "string length calculation" helper elsewhere in WIL that took into account the various forms of .get() / .size() / etc.

Reply:
we didn't find one - not that met the requirements of this, which separates out 3 distinct operations:
- find the length of the buffer within the string type
- pre-allocate a buffer within the string type of a specific byte size
- get a pointer to the string type buffer to have someone else write to it

Reply:
I think this is a really good idea. We should use the string_maker and extend it where needed.

Our decision for this PR was to implement our own thing for now. We're already at several thousand lines of code, and we didn't want to add even more complexity by mucking with the string helpers.

I've called this out as a follow-up point, since I agree that it's a good idea to converge our string manipulation and augment it where it's missing functionality. Resolving wontfix for now.